src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,163 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode;
+        if (size == 0) {
+            newNode = new Node<>(null, value, null);
+            head = newNode;
+        } else {
+            newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> nodeToAdd;
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("index: " + index + " Out of bound size : " + size);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> wantedNode = findNodeByIndex(index);
+        if (index == 0) {
+            nodeToAdd = new Node<>(null, value, wantedNode);
+            head = nodeToAdd;
+        } else {
+            nodeToAdd = new Node<>(wantedNode.prev, value, wantedNode);
+            wantedNode.prev.next = nodeToAdd;
+        }
+        wantedNode.prev = nodeToAdd;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> wantedNode = findNodeByIndex(index);
+        return wantedNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        Node<T> wantedNode = findNodeByIndex(index);
+        T oldValue = wantedNode.value;
+        wantedNode.value = value;
+        return oldValue;
+    }
+
+    private void indexCheck(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("index: " + index + " Out of bound size : " + size);
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> wantedNode = findNodeByIndex(index);
+        if (head == tail) {
+            head = null;
+            tail = null;
+        } else if (index == 0) {
+            head = wantedNode.next;
+            wantedNode.next.prev = null;
+        } else if (index == size - 1) {
+            tail = wantedNode.prev;
+            wantedNode.prev.next = null;
+        } else {
+            wantedNode.prev.next = wantedNode.next;
+            wantedNode.next.prev = wantedNode.prev;
+        }
+        unlink(wantedNode);
+        size--;
+        return wantedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        boolean result = false;
+        Node<T> wantedNode = head;
```

Comment:
```suggestion
        Node<T> node = head;
```
fix this naming everywhere